realm-jni/src/util.cpp

```diff
@@ -166,6 +166,16 @@ void jprint(JNIEnv *env, char *txt)
 #endif
 }
 
+void ThrowNullValueException(JNIEnv* env, Table* table, size_t col_ndx) {
+    std::ostringstream ss;
+    ss << "Trying to set a non-nullable field '"
+       << table->get_column_name(col_ndx)
+       << "' in '"
+       << table->get_name()
+       << "' to null.";
+    ThrowException(env, IllegalArgument, ss.str());
```

Comment:
What about throw NullPointerException instead? For boxed types, NullPotinerException will be send when cast null to boolean/long. For IllegalArgumentException, extra code needs to be written which can slow down the setter a little bit.